056-zk-kraft-migration.md

```diff
@@ -0,0 +1,451 @@
+The current reconciliation ends.
+
+**_Status changes_**
+
+The FSM moves back to the `WaitBrokersInMigration`, so actually at the beginning of the Phase 2.
+
+---
+
+If the user doesn't rollback the migration, next step is to finalize the migration process. This step is irreversible.
+
+**_Preconditions_**
+
+The FSM is in the `DualWrite` state (as per `Kafka.status.migrationState` field) and the operator expects the user desire to finalize the migration, having a full KRaft based cluster and ZooKeeper out of the picture.
+
+**_Trigger_**
+
+The user applies the `strimzi.io/kraft: post-migration` annotation to the `Kafka` custom resource.
+
+**_Actions_**
+
+The operator starts a reconciliation and, via the `KafkaReconciler` component, updates the brokers' configuration by removing the ZooKeeper migration flag and connection to the ZooKeeper ensemble.
+It also has to replace the `broker.id` with `node.id`.
+More specifically, the brokers are automatically updated with the following configuration (the user cannot set them manually).
+
+```shell
+# replace broker.id with node.id
+node.id=0
+
+# Remove the migration enabled flag
+# zookeeper.metadata.migration.enable=true
+
+# Remove ZooKeeper client configuration
+# zookeeper.connect=<connection-to-zookeeper>
+
+# Don't set the IBP, KRaft uses "metadata.version" feature flag
+# inter.broker.protocol.version=3.5
+```
+
+The operator rolls the brokers, via the `KafkaRoller`, and waits for them to be ready again.
+They are now connected to the KRaft controllers quorum only and not to the ZooKeeper ensemble anymore.
+
+When brokers are rolled and ready again, the operator updates the controllers' configuration by removing the ZooKeeper migration flag and connection to the ZooKeeper ensemble.
+More specifically, the controllers are automatically updated with the following configuration (the user cannot set them manually).
+
+```shell
+# Disable the migration
+# zookeeper.metadata.migration.enable=true
+
+# Remove ZooKeeper client configuration
+# zookeeper.connect=<connection-to-zookeeper>
+```
+
+Finally, the operator rolls the controllers, via the `KafkaRoller`, and waits for them to be ready again.
+They are now disconnected from the ZooKeeper ensemble which remain running but are no longer the source of truth for Kafka metadata.
+The current reconciliation ends.
+
+**_Status changes_**
+
+The FSM moves to the `FullKRaft` state and ZooKeeper is not used anymore.
+
+#### Phase 5: Removing ZooKeeper
+
+At this point the ZooKeeper ensemble is out of the picture, still running but not used anymore.
+
+**_Preconditions_**
+
+The FSM is in the `FullKRaft` state (as per `Kafka.status.migrationState` field) and the operator expects the user to finalize the migration by removing ZooKeeper.
+
+**_Trigger_**
+
+The user applies the `strimzi.io/kraft: post-migration` annotation and removes the `spec.zookeeper` on the `Kafka` custom resource.
```

Comment:
Should this be `enabled`? `post-migration` is already in the previous step as trigger.